File: Tayyab/Lab8_cocotb/restoring_division/sim/boothmultiplier_tb.cpp
/*
    Name: boothmultiplier_tb.cpp
    Author: Muhammad Tayyab
    Date: 6-8-2024
    Description: Verilator testbench for boothmultiplier.sv

    Comandline argument: argv[1]   : number of random tests to run
*/

#include <stdlib.h>
#include <iostream>
#include <verilated.h>
#include <verilated_vcd_c.h>
#include "Vbooth_multiplier.h"
#include "Vbooth_multiplier___024root.h"

#define NUMBITS 32

// Clocking variables
vluint64_t sim_time = 0;
vluint64_t clk_pedges = 0;

// Global variables
int num_rand_tests;
char monitor_state = 0;
signed int monitor_num_a, monitor_num_b;
signed long ref_result;
char driver_state = 0;
int num_tests = 0;
int passed_tests = 0, failed_tests = 0;

// Function prototypes
void driver(Vbooth_multiplier *dut);
void monitor(Vbooth_multiplier *dut);


int main(int argc, char** argv, char** env) {

    // Get command line arguments
    if (argc==2) {
        num_rand_tests = atoi(argv[1]);
    }
    else {
        num_rand_tests = 100;
    }

    // Connect to dut
    Vbooth_multiplier *dut = new Vbooth_multiplier;

    // Dump signals to wavefile
    Verilated::traceEverOn(true);
    VerilatedVcdC *m_trace = new VerilatedVcdC;
    dut->trace(m_trace, 5);
    m_trace->open("waveform.vcd");

    // Random seed
    srand(3456);

    // Loop until all random tests performed
    while (num_tests <= num_rand_tests) {
        dut->clk ^= 1;
        driver(dut);        // drive signals
        dut->eval();
        monitor(dut);       // monitor the outputs
        m_trace->dump(sim_time);
        sim_time++;
    }

    // Print results
    printf("\n##################\n");
    printf("Total random tests: %d\n", int(num_rand_tests));
    printf("Passed: %d\n", passed_tests);
    printf("Failed: %d\n", failed_tests);
    printf("\n##################\n");

    // Close wavefile and dut
    m_trace->close();
    delete dut;
    exit(EXIT_SUCCESS);
}

void driver(Vbooth_multiplier *dut) {
    // At negedge clk, hold values
    if (sim_time % 2 == 1) {
        dut->reset=dut->reset;
        dut->num_a=dut->num_a;
        dut->num_b=dut->num_b;
        dut->valid_src=dut->valid_src;
        dut->ready_dst=dut->ready_dst;
    }
    // At posedge clk, drive new values
    else {
        switch (driver_state)
        {
            // Init sequence
            case 0:
                dut->reset=1;
                dut->num_a=0;
                dut->num_b=0;
                dut->valid_src=0;
                dut->ready_dst=0;
                driver_state = 1;
                break;
            // Reset
            case 1 ... 3:
                dut->reset=0;
                dut->num_a=0;
                dut->num_b=0;
                dut->valid_src=0;
                dut->ready_dst=0;
                driver_state += 1;
                break;
            // Apply random inputs
            case 4:
                dut->reset=1;
                dut->num_a = rand();// % (1<<NUMBITS);
                dut->num_b = rand();// % (1<<NUMBITS);
                dut->valid_src=1;
                driver_state=5;
                num_tests += 1;
                break;
            // Wait for ready_src
            case 5:
                dut->reset=1;
                dut->num_a=dut->num_a;
                dut->num_b=dut->num_b;
                dut->valid_src=1;
                // if ready_src, deassert valid_src and go back to state 4
                if (dut->ready_src) {
                    dut->valid_src=0;
                    driver_state = 4;
                }
                break; 
            
            default:
                dut->reset=1;
                dut->num_a=0;
                dut->num_b=0;
                dut->valid_src=0;
                dut->ready_dst=0;
                driver_state = driver_state;
                break; 
        }
    clk_pedges += 1;
    }
}

void monitor(Vbooth_multiplier *dut) {
    // Don't run (return) on negedge clk
    if (sim_time % 2 == 1) {
        return;
    }
    
    // Wait for valid_src
    // Then sample inputs of dut 
    if (monitor_state==0 && dut->valid_src) {
        monitor_num_a = dut->num_a;
        monitor_num_b = dut->num_b;
        ref_result =  (long) monitor_num_a * monitor_num_b;
        monitor_state = 1;
        return;
    }

    // When valid_dst comes, check output of dut
    if (monitor_state==1 && dut->valid_dst) {
        long test_product = ((long) dut->product_high << NUMBITS) + dut->product_low;
        if (test_product != ref_result) {
            printf("\nTest fail.\n");
            failed_tests += 1;
        }
        else {
            passed_tests += 1;
        }
        monitor_state = 2;
        return;
    }

    // Assert ready_dst
    if (monitor_state==2) {
        dut->ready_dst = 1;
        monitor_state = 3;
        return;
    }

    // Deassert ready_dst
    if (monitor_state==3) {
        dut->ready_dst = 0;
        monitor_state = 0;
        return;
    }

}

File: Tayyab/Lab8_cocotb/booth_multiplier/sim/boothmultiplier_tb.py
"""
    Name: boothmultiplier_tb.sv
    Author: Muhammad Tayyab
    Date: 7-8-2024
    Description: Cocotb testbanch for booth_multiplier.sv
"""

import cocotb
from cocotb.triggers import Timer, RisingEdge, FallingEdge
from cocotb.clock import Clock
from random import randint
from ref_models.boothmultiplier import boothMultiplier as ref_model
import os
# Control vars
NUMBITS = 32
directed_test_pairs = [(-1,-1),(2,3),(3,-2),(-6,6)]
NUM_RAND_TESTS = int(1e3)

MASK = (2**NUMBITS)-1
TEST_BOUND = 2**(NUMBITS-1)
random_test_pairs = [(randint(-TEST_BOUND, TEST_BOUND-1), randint(-TEST_BOUND, TEST_BOUND-1)) for x in range(NUM_RAND_TESTS)]
directed_test_pairs.extend(random_test_pairs)

@cocotb.test()
async def directed_random_tests(dut):
    await init(dut)
    await reset_sequence(dut, 5)

    for test_pair in directed_test_pairs:
        await RisingEdge(clk)
        num_a.value = test_pair[0]
        num_b.value = test_pair[1]
        valid_src.value = 1

        ref_result = ref_model(test_pair[0], test_pair[1])
        ref_high_bits = ref_result[0] & MASK
        ref_low_bits = ref_result[1] & MASK

        while (not valid_dst.value):
            await RisingEdge(clk)
        
        assert (ref_high_bits, ref_low_bits) == (product_high.value, product_low.value)
        if 0:
            print("\n")
            print("a,b:",hex(num_a.value), hex(num_b.value))
            print("Product:",hex(product_high.value), hex(product_low.value))
            print("Ref    :",hex(ref_high_bits), hex(ref_low_bits))
            print("\n")

        ready_dst.value = 1
        await RisingEdge(clk)
        ready_dst.value = 0
        valid_src.value = 0
        

async def init(dut):
    # Global handles for signals
    global clk, reset, num_a, num_b, product_low, product_high
    global valid_src, ready_dst, ready_src, valid_dst
    
    clk=dut.clk
    reset=dut.reset
    num_a=dut.num_a
    num_b=dut.num_b
    product_high=dut.product_high
    product_low=dut.product_low
    valid_src=dut.valid_src
    ready_dst=dut.ready_dst
    ready_src=dut.ready_src
    valid_dst=dut.valid_dst

    # Initial values
    reset.value=1
    valid_src.value=0
    ready_dst.value=0

    # Start clock
    c = Clock(dut.clk, 10, 'step')
    await cocotb.start(c.start())

async def reset_sequence(dut, reset_time):
    reset.value = 0
    await Timer(reset_time)
    reset.value = 1

